mongo/datadog_checks/mongo/dbm/operation_samples.py

```diff
@@ -0,0 +1,359 @@
+# (C) Datadog, Inc. 2024-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+
+
+import time
+from typing import Dict, List, Optional
+
+from bson import json_util
+
+try:
+    import datadog_agent
+except ImportError:
+    from datadog_checks.base.stubs import datadog_agent
+
+from datadog_checks.base.utils.db.sql import compute_exec_plan_signature
+from datadog_checks.base.utils.db.utils import DBMAsyncJob
+from datadog_checks.base.utils.tracking import tracked_method
+from datadog_checks.mongo.common import ReplicaSetDeployment
+
+from .types import (
+    OperationActivityEvent,
+    OperationSampleActivityRecord,
+    OperationSampleClient,
+    OperationSampleEvent,
+    OperationSampleOperationMetadata,
+    OperationSampleOperationStats,
+    OperationSamplePlan,
+)
+
+
+class MongoDbExplainExceptionBase(Exception):
+    pass
+
+
+SAMPLE_EXCLUDE_KEYS = {
+    'statement',
+    'application',
+    'dbname',
+    'user',
+    'client',
+}
+
+
+def agent_check_getter(self):
+    return self._check
+
+
+class MongoOperationSamples(DBMAsyncJob):
+    def __init__(self, check):
+        self._operation_samples_config = check._config.operation_samples
+        self._collection_interval = self._operation_samples_config['collection_interval']
+
+        super(MongoOperationSamples, self).__init__(
+            check,
+            rate_limit=1 / self._collection_interval,
+            run_sync=not self._operation_samples_config.get('async', False),  # Default to running sync
+            enabled=self._operation_samples_config['enabled'],
+            dbms="mongodb",
+            min_collection_interval=check._config.min_collection_interval,
+            job_name="operation-samples",
+        )
+
+    def run_job(self):
+        self.collect_operation_samples()
+
+    @tracked_method(agent_check_getter=agent_check_getter)
+    def collect_operation_samples(self):
+        if not self._should_collect_operation_samples():
+            return
+
+        now = time.time()
+
+        activities = []
+
+        for sample in self._get_operation_samples(now, activities):
+            self._check.log.debug("Sending operation sample: %s", sample)
+            # Emit operation sample event
+            self._check.database_monitoring_query_sample(json_util.dumps(sample))
+
+        activities_payload = self._create_activities_payload(now, activities)
+        self._check.log.debug("Sending activities payload: %s", activities_payload)
+        # Emit activities event
+        self._check.database_monitoring_query_activity(json_util.dumps(activities_payload))
```

Comment:
This flow is weird to me and looks bug-prone. You're passing in a mutable list `activities`, modifying that list, but also returning each item of the list? Is there a reason you've structured it this way (e.g. memory?)

I think a cleaner flow could be:


```python
activites = self._get_operation_samples(now, activities)

for sample in activities:
    explain_plan = self._get_explain_plan(sample)
    sample['explain_plan'] = explain_plan
    self._check.database_monitoring_query_sample(json_util.dumps(sample))

activities_payload = self._create_activities_payload(now, activities)
self._check.database_monitoring_query_activity(json_util.dumps(activities_payload))
```

Reply:
The `_get_operation_samples` internally uses `$currentOp` aggregation pipeline, which returns a stream of documents for each sampled operation as compared to one big ducoment for all operations. Inside the function, it reformat every operation into an activity row (drops unneeded fields) and append it to the `activities` list. At the end of the function, it also yield the sample with explain plan. This way we only keep one sample in the memory to submit as sample event, and a list of activities to be submitted at the end of the process.
If we were to convert the all sampled operations into `activities` at the beginning, we will keep 2 copies of every operation. 
However I do agree passing a mutable `activities` list looks weird and not optimal. To make the flow clear, we can build `activities` list from samples to avoid passing around mutable list.